### PR TITLE
[#160521227] Add withLoadingSpinner HOC

### PR DESCRIPTION
**Pull request by fpersico**
/**
 * A HOC to display and overlay spinner conditionally
 *
 * @param WrappedComponent The react component you want to wrap
 * @param isLoadingSelector A redux selector that returns true when the spinner need to be visible
 * @param spinnerProps Props to pass to the spinner component
 */

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Remember to fix the PR title by removing WIP wording when ready
  </td>
    </tr>
  </tbody>
</table>



## Affected stories

  * 🌟 [#160521227](https://www.pivotaltracker.com/story/show/160521227): L'app deve mostrare uno spinner di caricamento quando vengono effettuate azioni bloccanti durante un pagamento


New dependencies added: <a href='https://www.npmjs.com/package/hoist-non-react-statics'><code>hoist-non-react-statics</code></a>, <a href='https://www.npmjs.com/package/react-native-loading-spinner-overlay'><code>react-native-loading-spinner-overlay</code></a>, <a href='https://www.npmjs.com/package/@types/hoist-non-react-statics'><code>&#64;types/hoist-non-react-statics</code></a> and <a href='https://www.npmjs.com/package/@types/react-native-loading-spinner-overlay'><code>&#64;types/react-native-loading-spinner-overlay</code></a>.


<h2><a href="https://github.com/mridgway/hoist-non-react-statics#readme">hoist-non-react-statics</a></h2>
<p>Author: Michael Ridgway</p>
<p>Description: Copies non-react specific statics from a child component to a parent component</p>
<p>Homepage: <a href="https://github.com/mridgway/hoist-non-react-statics#readme">https://github.com/mridgway/hoist-non-react-statics#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 3 years ago</td></tr><tr><td>Last Updated</td><td>about 2 months ago</td></tr><tr><td>License</td><td>BSD-3-Clause</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>27</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/react-is'><code>react-is</code></a></td></tr><tr><td>Keywords</td><td>react</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# hoist-non-react-statics

[NPM version](http://badge.fury.io/js/hoist-non-react-statics)
[Build Status](https://travis-ci.org/mridgway/hoist-non-react-statics)
[Coverage Status](https://coveralls.io/r/mridgway/hoist-non-react-statics?branch=master)
[Dependency Status](https://david-dm.org/mridgway/hoist-non-react-statics)
[devDependency Status](https://david-dm.org/mridgway/hoist-non-react-statics#info=devDependencies)

Copies non-react specific statics from a child component to a parent component. 
Similar to `Object.assign`, but with React static keywords blacklisted from
being overridden.

```bash
$ npm install --save hoist-non-react-statics
```

## Usage

```js
import hoistNonReactStatics from 'hoist-non-react-statics';

hoistNonReactStatics(targetComponent, sourceComponent);
```

If you have specific statics that you don't want to be hoisted, you can also pass a third parameter to exclude them:

```js
hoistNonReactStatics(targetComponent, sourceComponent, { myStatic: true, myOtherStatic: true });
```

## What does this module do?

See this [explanation](https://facebook.github.io/react/docs/higher-order-components.html#static-methods-must-be-copied-over) from the React docs.

## Compatible React Versions

Please use latest 3.x unless you need to support react@0.13. Versions prior to 3.x will not support ForwardRefs.

| hoist-non-react-statics Version | Compatible React Version |
|--------------------------|-------------------------------|
| 3.x | 0.14-16.x With ForwardRef Support |
| 2.x | 0.13-16.x Without ForwardRef Support |
| 1.x | 0.13-16.2 |

## Browser Support

This package uses `Object.defineProperty` which has a broken implementation in IE8. In order to use this package in IE8, you will need a polyfill that fixes this method.

## License
This software is free to use under the Yahoo Inc. BSD license.
See the [LICENSE file][] for license text and copyright information.

[LICENSE file]: https://github.com/mridgway/hoist-non-react-statics/blob/master/LICENSE.md

Third-party open source code used are listed in our [package.json file]( https://github.com/mridgway/hoist-non-react-statics/blob/master/package.json).


</details>




<h2><a href="https://github.com/joinspontaneous/react-native-loading-spinner-overlay">react-native-loading-spinner-overlay</a></h2>
<p>Author: Nick Baugh</p>
<p>Description: The only pure React Native Native iOS and Android loading spinner (progress bar indicator) overlay</p>
<p>Homepage: <a href="https://github.com/joinspontaneous/react-native-loading-spinner-overlay">https://github.com/joinspontaneous/react-native-loading-spinner-overlay</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 2 years ago</td></tr><tr><td>Last Updated</td><td>8 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>20</td></tr><tr><td>Keywords</td><td>activity, android, bar, full, indicator, ios, loading, modal, native, pop-up, popup, progress, react, react-native, screen, spinner and view</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>


# React Native Loading Spinner Overlay

[![NPM version][npm-image]][npm-url]
[![NPM downloads][npm-downloads]][npm-url]
[![MIT License][license-image]][license-url]

> **tldr;** The only pure [React Native][react-native] Native iOS and Android loading spinner (progress bar indicator) overlay

![Demo][demo]


## Index

* [Install](#install)
* [Usage](#usage)
* [Platforms](#platforms)
* [Notes](#notes)
* [Development](#development)
* [Contributors](#contributors)
* [Credits](#credits)
* [License](#license)


## Install

For React Native version `>=0.28.x` use version `>=0.3.x` (`0.2.x` is broken, sorry!):

```bash
npm install --save react-native-loading-spinner-overlay@latest
```

For React Native version `<=0.27.x` use version `0.1.x`:

```bash
npm install --save react-native-loading-spinner-overlay@0.1.x
```


## Usage

This usage shows the default styles and properties.

| Property      | Type           |   Default  | Description  |
|---------------|----------------|------------|--------------|
| cancelable    | `boolean`      |    `false` | **Android**: If set to false, it will prevent spinner from hiding when pressing the hardware back button.  If set to true, it will allow spinner to hide if the hardware back button is pressed. |
| color         | `string`      |   `white` | Changes the spinner's color (example values are `red`, `#ff0000`, etc). For adjusting the contrast see `overlayColor` prop below.|
| animation     |`none`, `slide`, `fade`|   `none`  | Changes animation on show and hide spinner's view.|
| overlayColor  | `string`      |    `rgba(0, 0, 0, 0.25)` | Changes the color of the overlay.|
| size          | `small`, `normal`, `large` | `large ` | Sets the spinner's size. No other cross-platform sizes are supported right now.|
| textContent  | `string`      |    `""` | Optional text field to be shown.|
| textStyle  | `style`      |    `-` | The style to be applied to the `<Text>` that displays the `textContent`.|
| visible  | `boolean`      |    `false` | Controls the visibility of the spinner.|

You can also add a child view to act as a custom activity indicator.

```js
import React, { View, Text } from 'react-native';

import Spinner from 'react-native-loading-spinner-overlay';

class MyComponent extends React.Component {

  constructor(props) {
    super();
    this.state = {
      visible: false
    };
  }

  /* eslint react/no-did-mount-set-state: 0 */
  componentDidMount() {
    setInterval(() => {
      this.setState({
        visible: !this.state.visible
      });
    }, 3000);
  }

  render() {
    return (
      <View style={{ flex: 1 }}>
        <Spinner visible={this.state.visible} textContent={"Loading..."} textStyle={{color: '#FFF'}} />
      </View>
    );
  }
}
```

To use a custom activity indicator just pass it as child of the component:

```js
<Spinner visible={this.state.visible}>
  <Text>This is my custom spinner</Text>
</Spinner>
```

## Platforms

> For `>= 0.3.x`:

* We use `ActivityIndicator` now!

> For `0.2.x`:

* Do not use this version due to [#22][issue-22], use `>= 0.3.x` please!

> For `<= 0.1.x`:

* iOS: this platform uses `Modal` ([docs][modal-docs]/[source][modal-source]) to overlay and `ActivityIndicatorIOS` ([docs][activity-indicator-docs]) for the loading spinner
* Android: this platform uses `Portal` ([source][portal-source]) to overlay and `ActivityIndicator` ([docs][progress-bar-docs]) for the loading spinner


## Notes

> For `>= 0.3.x`:

* We use `ActivityIndicator` now!

> For `0.2.x`:

* This version is broken due to a dependency issue, see issue [#22][issue-22]

> For `<= 0.1.x`:

* Docs don't exist yet for `Portal`, see [this issue on GitHub][portal-docs-issue]; once those are in, then we can add a link to the source in [Platforms](#platforms)
* Until a release of React Native is shipped [for this pull request][style-attr-issue], Android's `ProgressBarAndroid` will not have support for a `StyleAttr` value of `"Normal"` - therefore we only support a `size` prop of `"small"` or `"large"` right now (defaulting to `"large"`) - in other words, we can only support Android's inverse styling with a `styleAttr` of `"Inverse"`, `"SmallInverse"` (for a `size` prop of `"small"`), and `"LargeInverse"` (for a `size` prop of `"large`") (since there is no `"Normal"` support right now for `"size"` of `"normal"`).


## Development

1. Fork/clone this repository
2. Run `npm install`
4. Make changes in `src` directory
6. Run `npm test` when you're done
7. Submit a pull request


## Contributors

* Nick Baugh <niftylettuce@gmail.com>


## License

[MIT][license-url]


[license-image]: http://img.shields.io/badge/license-MIT-blue.svg?style=flat
[license-url]: LICENSE
[react-native]: https://facebook.github.io/react-native
[npm-image]: http://img.shields.io/npm/v/react-native-loading-spinner-overlay.svg?style=flat
[npm-url]: https://npmjs.org/package/react-native-loading-spinner-overlay
[npm-downloads]: http://img.shields.io/npm/dm/react-native-loading-spinner-overlay.svg?style=flat
[portal-source]: https://github.com/facebook/react-native/blob/master/Libraries/Portal/Portal.js
[modal-docs]: https://facebook.github.io/react-native/docs/modal.html
[modal-source]: https://github.com/facebook/react-native/blob/master/Libraries/Modal/Modal.js
[demo]: https://cdn.rawgit.com/niftylettuce/react-native-loading-spinner-overlay/master/media/demo.gif
[activity-indicator-docs]: https://facebook.github.io/react-native/docs/activityindicatorios.html
[progress-bar-docs]: https://facebook.github.io/react-native/docs/activityindicator.html
[portal-docs-issue]: https://github.com/facebook/react-native/issues/2501
[style-attr-issue]: https://github.com/facebook/react-native/pull/4974
[issue-22]: https://github.com/niftylettuce/react-native-loading-spinner-overlay/issues/22


</details>




<h2><a href="http://npmjs.com/package/@types/hoist-non-react-statics">&#64;types/hoist-non-react-statics</a></h2>
<p>Author: Unknown</p>
<p>Description: TypeScript definitions for hoist-non-react-statics</p>
<p>Homepage: <a href="http://npmjs.com/package/@types/hoist-non-react-statics">http://npmjs.com/package/@types/hoist-non-react-statics</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>about 2 months ago</td></tr><tr><td>Last Updated</td><td>about 2 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>2</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/@types/react'><code>&#64;types/react</code></a></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Installation
> `npm install --save @types/hoist-non-react-statics`

# Summary
This package contains type definitions for hoist-non-react-statics (https://github.com/mridgway/hoist-non-react-statics#readme).

# Details
Files were exported from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/hoist-non-react-statics

Additional Details
 * Last updated: Thu, 02 Aug 2018 01:35:11 GMT
 * Dependencies: react
 * Global values: none

# Credits
These definitions were written by JounQin <https://github.com/JounQin>.


</details>




<h2><a href="http://npmjs.com/package/@types/react-native-loading-spinner-overlay">&#64;types/react-native-loading-spinner-overlay</a></h2>
<p>Author: Unknown</p>
<p>Description: TypeScript definitions for react-native-loading-spinner-overlay</p>
<p>Homepage: <a href="http://npmjs.com/package/@types/react-native-loading-spinner-overlay">http://npmjs.com/package/@types/react-native-loading-spinner-overlay</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>7 months ago</td></tr><tr><td>Last Updated</td><td>about 1 month ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>2</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/@types/react'><code>&#64;types/react</code></a> and <a href='https://www.npmjs.com/package/@types/react-native'><code>&#64;types/react-native</code></a></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Installation
> `npm install --save @types/react-native-loading-spinner-overlay`

# Summary
This package contains type definitions for react-native-loading-spinner-overlay (https://github.com/joinspontaneous/react-native-loading-spinner-overlay).

# Details
Files were exported from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react-native-loading-spinner-overlay

Additional Details
 * Last updated: Mon, 06 Aug 2018 21:55:27 GMT
 * Dependencies: react, react-native
 * Global values: none

# Credits
These definitions were written by fhelwanger <https://github.com/fhelwanger>.


</details>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by cloudify** [comment]
I think we need to add some text that explains what's going on - the use has to (potentially) wait for several seconds and we need to tell him to wait

**Comment by fpersico** [comment]
spinnerProps has textContent to display the message

export interface SpinnerProps {
    cancelable?: boolean;
    color?: string;
    animation?: "none" | "slide" | "fade";
    overlayColor?: string;
    size?: "normal" | "small" | "large";
    textContent?: string;
    textStyle?: ReactNative.StyleProp<ReactNative.TextStyle>;
    visible?: boolean;
}

example usage

withLoadingSpinner(Mycomponent, myIsLoadingSelector, { textContent: "I am loading data please wait" })

File: ts/components/helpers/withLoadingSpinner.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+};
+
+/**
+ * A HOC to display and overlay spinner conditionally
+ *
+ * @param WrappedComponent The react component you want to wrap
+ * @param isLoadingSelector A redux selector that returns true when the spinner need to be visible
+ * @param spinnerProps Props to pass to the spinner component
+ */
+export function withLoadingSpinner<P>(
```
Comment (cloudify) [inline ts/components/helpers/withLoadingSpinner.tsx]: no components with `State` ?

Comment (fpersico) [reply ts/components/helpers/withLoadingSpinner.tsx]: What you exactly mean? 

```
export function withLoadingSpinner<P, S>
```

i think we don't need it because

```
WrappedComponent: React.ComponentType<P>
```

ComponentType only needs P

Comment (cloudify) [reply ts/components/helpers/withLoadingSpinner.tsx]: what I mean is that this HOC discards the `State` type, but perhaps that's by design since `State` is not exposed outside of the component.

Comment (fpersico) [reply ts/components/helpers/withLoadingSpinner.tsx]: Yes exactly.
